node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -1001,57 +1012,72 @@ pub(crate) async fn run<Context>(
 	futures::pin_mut!(next_inactivity_stream);
 
 	loop {
-		let res = {
-			let s = futures::future::select(ctx.recv().fuse(), next_inactivity_stream.next().fuse());
-
-			if let Poll::Ready(res) = futures::poll!(s) {
-				Some(match res {
-					Either::Left((msg, _)) => Either::Left(msg.map_err(Fatal::SubsystemReceive)?),
-					Either::Right((_, _)) => Either::Right(()),
-				})
-			} else {
-				None
-			}
-		};
-
-		match res {
-			Some(Either::Left(msg)) => {
-				tracing::trace!(target: LOG_TARGET, msg = ?msg, "received a message");
-
-				match msg {
-					Communication { msg } => process_msg(
-						&mut ctx,
-						&keystore,
-						msg,
-						&mut state,
-					).await,
-					Signal(BlockFinalized(..)) => {}
-					Signal(ActiveLeaves(_)) => {}
-					Signal(Conclude) => { break }
+		select! {
+			res = ctx.recv().fuse() => {
+				match res {
+					Ok(Communication { msg }) => {
+						tracing::trace!(target: LOG_TARGET, msg = ?msg, "received a message");
+						process_msg(
+							&mut ctx,
+							&keystore,
+							msg,
+							&mut state,
+						).await;
+					}
+					Ok(Signal(Conclude)) => break,
+					_ => {},
 				}
-
-				continue
 			}
-			Some(Either::Right(())) => {
+			_ = next_inactivity_stream.next() => {
 				disconnect_inactive_peers(&mut ctx, &eviction_policy, &state.peer_data).await;
-				continue
 			}
-			None => {}
+			res = state.collations.next() => {
+				// If no prior collation for this relay parent has been seconded, then
+				// memoize the collation_event for that relay_parent, such that we may
+				// notify the collator of their successful second backing
+				if let Some((relay_parent, collation_event)) = if let Some(
+					(collation_event, Some(Ok((candidate_receipt, pov))))
```

Comment:
if collation fetching times out, we will not insert into `state.seconded`, right? do we want to log that?